Sources/TuistCache/ContentHashing/TestsGraphContentHasher.swift

```diff
@@ -0,0 +1,69 @@
+import Foundation
+import TSCBasic
+import TuistCore
+import TuistGraph
+import TuistSupport
+
+protocol TestsGraphContentHashing {
+    func contentHashes(graph: Graph) throws -> [TargetNode: String]
+}
+
+/// `TestsGraphContentHasher`
+/// is responsible for computing an hash that uniquely identifies a Tuist `Graph`.
+/// It hashes only unit tests and its dependencies
+final class TestsGraphContentHasher: TestsGraphContentHashing {
+    private let targetContentHasher: TargetContentHashing
+
+    // MARK: - Init
+
+    init(
+        targetContentHasher: TargetContentHashing = TargetContentHasher(
+            contentHasher: ContentHasher()
+        )
+    ) {
+        self.targetContentHasher = targetContentHasher
+    }
+
+    // MARK: - TestsGraphContentHashing
+
+    func contentHashes(graph: Graph) throws -> [TargetNode: String] {
+        var visitedTargets: [TargetNode: Bool] = [:]
+        let hashableTargets = graph.targets
+            .flatMap(\.value)
+            // UI tests depend on the device they are run on
+            // This can be done in the future if we hash the ID of the device
+            // But currently, we consider for hashing only unit tests and its dependencies
+            .filter { $0.target.product == .unitTests }
+            .flatMap { target -> [TargetNode] in
+                targetDependencies(
+                    target,
+                    visited: &visitedTargets
+                )
+            }
+        let hashes = try hashableTargets.map {
+            try targetContentHasher.contentHash(
+                for: $0,
+                cacheOutputType: .none
```

Comment:
I'd recommend changing the logic of the hashers to be cache agnostic. That'd mean removing this argument and also the internal checks that check whether the target is a framework or not. What you can do though to speed things up since you don't want to hash all the targets is passing a function that takes a target as an input and returns a boolean indicating whether the target should be hashed or not.

Reply:
> I'd recommend changing the logic of the hashers to be cache agnostic.

do you mean to strip `cacheOutputType` from `TargetContentHasher`? And for the `tuist cache` rehashing the output from `TargetContentHasher` with the `cacheOutputType` which is necessary to differentiate between `.xcframework` and `.framework`?

Or do you mean to merge `TestsGraphContentHasher` and `GraphContentHasher`? The logic for those is a little bit different and passing a simple function will not do. What I'd have to do is go through unit tests target and their dependencies, save them into array, and then in the passed function check if the target is in that array.

But I am not sure if that approach is very efficient + we actually would not reuse a lot of logic anyway, so I am not sure if such approach is worth it (I'd prefer just removing `cacheOutputType` from `TargetContentHasher`)

> also the internal checks that check whether the target is a framework or not.

do you mean checks in `GraphContentHasher`? Because there are no such checks for `TestsGraphContentHasher` and `TargetContentHasher`.

Reply:
What I mean is that I'd modify the hasher to have no knowledge on the cache. If you look at the implementation, internally we check if the target is cacheable in order to hash it. In order to make it reusable, I think we can remove that piece, and rather pass a function that allows filtering. That way we can control from what targets we'd like to hash. Like in your case, you are only interested in hashing the targets that are impacted by the selective testing. 
I imagine an API along the lines of:

```swift
func hash(graph: ValueGraph, take: (ValueGraphTarget) -> Bool) -> [ValueGraphTarget: String]
```

Notice the `take` attribute that allows the consumer of the API to choose the targets.

Reply:
@pepibumur are you OK with creating `CacheGraphContentHasher` so the logic of filtering and hashing different output types could be reused? It will leverage `GraphContentHasher`, which will be cache agnostic.

I'm rather asking in advance because this change means a little more of moving around files and tests, so I don't want to waste that effort.

Reply:
> @pepibumur are you OK with creating CacheGraphContentHasher so the logic of filtering and hashing different output types could be reused? It will leverage GraphContentHasher, which will be cache agnostic.

I'm not sure if I follow this one.

Reply:
I re-read it again and I understood what you mean. I think it's good to have that `CacheGraphContentHasher` for now.